### ISSUE TITLE
解决图片输出全黑问题

### ISSUE THREAD

**Issue by klyx9** [SSinyu/WGAN_VGG#17]
1.**我无法获取原始的数据集，但是根据test函数中的反归一化操作标明输入的数据默认已经完成归一化。
我的数据格式为:
data_path
   ├── noise_image
   │       ├── 1.png
   │       ├── 2.png
   │       └── ...
   └──clean_image
           ├──1.png
           ├──2.png
在loader.py中对加载的数据进行归一化操作：**

class ct_dataset(Dataset):
    def __init__(self, mode, load_mode, saved_path, test_patient, patch_n=None, patch_size=None, transform=None):
        assert mode in ['train', 'test'], "mode is 'train' or 'test'"
        assert load_mode in [0,1], "load_mode is 0 or 1"

        if mode == "train":
            self.input_path = r".data_path/train/noise_image"
            self.target_path = r".data_path/train/clean_image"
        else:
            self.input_path = r".data_path/val/noise_image"
            self.target_path = r".data_path/val/clean_image"

        self.load_mode = load_mode
        self.patch_n = patch_n
        self.patch_size = patch_size
        self.transform = transform

        if mode == 'train':
            self.input_ = os.listdir(self.input_path)
            self.target_ = os.listdir(self.target_path)
        else: # mode =='test'
            self.input_ = os.listdir(self.input_path)
            self.target_ = os.listdir(self.target_path)

    def __len__(self):
        return len(self.target_)

    def __getitem__(self, idx):
        target_img_idx = self.target_[idx]
        input_img_path = os.path.join(self.input_path, target_img_idx)
        target_img_path = os.path.join(self.target_path, target_img_idx)

        input_img = np.array(Image.open(input_img_path).convert('L'))
        target_img = np.array(Image.open(target_img_path).convert('L'))

        input_img = input_img / 255  # 归一化
        target_img = target_img / 255  # 归一化

        if self.transform:
            input_img = self.transform(input_img)
            target_img = self.transform(target_img)

        if self.patch_size:
            input_patches, target_patches = get_patch(input_img,
                                                      target_img,
                                                      self.patch_n,
                                                      self.patch_size)
            return (input_patches, target_patches)
        else:
            return (input_img, target_img)

def get_patch(full_input_img, full_target_img, patch_n, patch_size, drop_background=0.1):
    assert full_input_img.shape == full_target_img.shape
    patch_input_imgs = []
    patch_target_imgs = []
    h, w = full_input_img.shape
    new_h, new_w = patch_size, patch_size
    n = 0
    while n <= patch_n:
        top = np.random.randint(0, h - new_h)
        left = np.random.randint(0, w - new_w)
        patch_input_img = full_input_img[top:top + new_h, left:left + new_w]
        patch_target_img = full_target_img[top:top + new_h, left:left + new_w]

        if (np.mean(patch_input_img) < drop_background) or \
            (np.mean(patch_target_img) < drop_background):
            continue
        else:
            n += 1
            patch_input_imgs.append(patch_input_img)
            patch_target_imgs.append(patch_target_img)
    '''
    for _ in range(patch_n):
        top = np.random.randint(0, h-new_h)
        left = np.random.randint(0, w-new_w)
        patch_input_img = full_input_img[top:top+new_h, left:left+new_w]
        patch_target_img = full_target_img[top:top+new_h, left:left+new_w]
        patch_input_imgs.append(patch_input_img)
        patch_target_imgs.append(patch_target_img)
    '''
    return np.array(patch_input_imgs), np.array(patch_target_imgs)

def get_loader(mode='train', load_mode=0,
               saved_path=None, test_patient='L506',
               patch_n=None, patch_size=None,
               transform=None, batch_size=32, num_workers=6):
    dataset_ = ct_dataset(mode, load_mode, saved_path, test_patient, patch_n, patch_size, transform)
    data_loader = DataLoader(dataset=dataset_, batch_size=batch_size, shuffle=True, num_workers=num_workers)
    return data_loader

2.**满足归一化后，直接训练还是全黑，训练就读不稳定，随时梯度爆炸或梯度消失，此时我参考了mmagic套件中的WGAN_GP代码，链接为：https://github.com/open-mmlab/mmagic/tree/main/configs/wgan-gp，使用他们的网络利用随机噪声进行图片过程发现训练稳定，在此基础上我查看了他们的生成器与判别器，发现大量使用BN归一化操作，随后我将BN层加入鉴别器和生成器网络中。**

class WGAN_VGG_generator(nn.Module):
    def __init__(self):
        super(WGAN_VGG_generator, self).__init__()
        layers = [nn.Conv2d(1,32,3,1,1), nn.BatchNorm2d(32), nn.ReLU(inplace=True),]
        for i in range(2, 8):
            layers.append(nn.Conv2d(32,32,3,1,1))
            layers.append(nn.BatchNorm2d(32))
            layers.append(nn.ReLU(inplace=True))
        layers.extend([nn.Conv2d(32,1,3,1,1), nn.Tanh()])
        self.net = nn.Sequential(*layers)

    def forward(self, x):
        out = self.net(x)
        return out + x

class WGAN_VGG_discriminator(nn.Module):
    def __init__(self, input_size):
        super(WGAN_VGG_discriminator, self).__init__()
        def conv_output_size(input_size, kernel_size_list, stride_list):
            n = (input_size - kernel_size_list[0]) // stride_list[0] + 1
            for k, s in zip(kernel_size_list[1:], stride_list[1:]):
                n = (n - k) // s + 1
            return n

        def add_block(layers, ch_in, ch_out, stride):
            layers.append(nn.Conv2d(ch_in, ch_out, 3, stride, 0))
            layers.append(nn.LeakyReLU(negative_slope=0.2, inplace=True))
            layers.append(nn.GroupNorm(num_groups=1, num_channels=ch_out, affine=True))
            return layers

        layers = []
        ch_stride_set = [(1,64,1),(64,64,2),(64,128,1),(128,128,2),(128,256,1),(256,256,2)]
        for ch_in, ch_out, stride in ch_stride_set:
            add_block(layers, ch_in, ch_out, stride)

        self.output_size = conv_output_size(input_size, [3]*6, [1,2]*3)
        self.net = nn.Sequential(*layers)
        self.fc1 = nn.Linear(256*self.output_size*self.output_size, 1024)
        self.fc2 = nn.Linear(1024, 1)
        self.lrelu = nn.LeakyReLU()

    def forward(self, x):
        out = self.net(x)
        out = out.view(-1, 256*self.output_size*self.output_size)
        out = self.lrelu(self.fc1(out))
        out = self.fc2(out)
        return out

加入BN层后训练稳定，输出的特征图逐渐趋向LDCT，解决黑屏问题。

3.可视化
**为了更好的观察生成器生成图片的过程，我训练过程阶段性的保存了部分生成图片。**
class WGAN_VGG(nn.Module):
    # referred from https://github.com/kuc2477/pytorch-wgan-gp
    def __init__(self, input_size=64):
        super(WGAN_VGG, self).__init__()
        self.generator = WGAN_VGG_generator()
        self.discriminator = WGAN_VGG_discriminator(input_size)
        self.feature_extractor = WGAN_VGG_FeatureExtractor()
        self.p_criterion = nn.L1Loss()
        self.sign = 0

    def d_loss(self, x, y, gp=True, return_gp=False):
        fake = self.generator(x).detach()
        d_real = self.discriminator(y)
        d_fake = self.discriminator(fake)
        d_loss = -torch.mean(d_real) + torch.mean(d_fake)
        if gp:
            gp_loss = self.gp(y, fake)
            loss = d_loss + gp_loss
        else:
            gp_loss = None
            loss = d_loss
        return (loss, gp_loss) if return_gp else loss

    def g_loss(self, x, y, perceptual=True, return_p=False):
        fake = self.generator(x)
        d_fake = self.discriminator(fake)
        g_loss = -torch.mean(d_fake)
        if perceptual:
            p_loss = self.p_loss(x, y)
            loss = g_loss + (0.1 * p_loss)
        else:
            p_loss = None
            loss = g_loss

        _a = fake.cpu().detach()
        _b = y.cpu().detach()

        if self.sign % 100 == 0:
            for num in range(80):
                a = _a[num].squeeze()
                a = a.numpy()
                a = (a * 255).astype(np.uint8)
                cv2.imwrite(rf"./save/{num}_fake.png", a)

                b = _b[num].squeeze()
                b = b.numpy()
                b = (b * 255).astype(np.uint8)
                cv2.imwrite(rf"./save/{num}_real.png", b)
        self.sign = self.sign + 1

        return (loss, p_loss) if return_p else loss

    def p_loss(self, x, y):
        fake = self.generator(x).repeat(1,3,1,1)
        real = y.repeat(1,3,1,1)
        fake_feature = self.feature_extractor(fake)
        real_feature = self.feature_extractor(real)
        loss = self.p_criterion(fake_feature, real_feature)
        return loss

    def gp(self, y, fake, lambda_=10):
        assert y.size() == fake.size()
        a = torch.cuda.FloatTensor(np.random.random((y.size(0), 1, 1, 1)))
        interp = (a*y + ((1-a)*fake)).requires_grad_(True)
        d_interp = self.discriminator(interp)
        fake_ = torch.cuda.FloatTensor(y.shape[0], 1).fill_(1.0).requires_grad_(False)
        gradients = torch.autograd.grad(
            outputs=d_interp, inputs=interp, grad_outputs=fake_,
            create_graph=True, retain_graph=True, only_inputs=True
        )[0]
        gradients = gradients.view(gradients.size(0), -1)
        gradient_penalty = ((gradients.norm(2, dim=1) -1)**2).mean() * lambda_
        return gradient_penalty

4.部分结果展示

![Image](https://github.com/user-attachments/assets/6c15663c-48fc-44ea-befb-06cd1859dcbe)
生成图片

![Image](https://github.com/user-attachments/assets/5c35d8ca-fcea-4f93-93bc-620f3b37ad7e)
LDCT

![Image](https://github.com/user-attachments/assets/75bad299-42e1-4f67-99ac-7bc27596b926)
HDCT

**Comment by XePWang** [comment]
我也遇到了这个问题，在这里我给出我的解决方法吧，在networks.py中
**存在问题的代码**（修复前的`networks.py:16`）：
```python
layers.extend([nn.Conv2d(32,1,3,1,1), nn.ReLU()])  # 最后一层有ReLU
self.net = nn.Sequential(*layers)
```

我修改成了
**修正后代码**：
```python
layers = [nn.Conv2d(1,32,3,1,1), nn.ReLU()]
for i in range(2, 8):
    layers.append(nn.Conv2d(32,32,3,1,1))
    layers.append(nn.ReLU())
layers.append(nn.Conv2d(32,1,3,1,1))  # 最后一层无激活函数
self.net = nn.Sequential(*layers)
```
你可以试试看，我这边这样已经可以解决这个输出全黑的问题了，不过模型性能比较一般
